Initial port for wasmJs

eygraber opened the pull request:
I'd like to supply a common interface, but at first glance that looks like it might be a lot more work.

It seems like there is partial support for null keys in js, but [the spec](https://www.w3.org/TR/IndexedDB/#key-construct) seems to suggest that it shouldn't be allowed. I mirrored the null key support in wasmJs for now.

Fixes #157

cedrickcooke commented:
This is awesome, thank you! I'll start looking at it ASAP

cedrickcooke commented on external/src/commonMain/kotlin/Interop.kt:
```diff
@@ -0,0 +1,73 @@
+@PublishedApi
+internal expect fun Array<IDBKey>.unsafeCast(): IDBKey
+
+public expect interface JsAny
+public expect interface UnsafeJsAny : JsAny
```
What's the difference between `JsAny` and `UnsafeJsAny`?

cedrickcooke commented on external/src/commonMain/kotlin/Interop.kt:
```diff
@@ -0,0 +1,73 @@
+@PublishedApi
+internal expect fun Array<IDBKey>.unsafeCast(): IDBKey
+
+public expect interface JsAny
```
Would love to see a doc comment on this explaining it. From my understanding, `JsAny` is meant as a marker of DB-safe types. Ideally something that includes the`external interface MyRowType : JsAny` use case.

eygraber replied:
`JsAny` is really just there so that there can be common code for Js and Wasm. Wasm requires externals to extend `JsAny`, so that had to be represented in the common code. 

I believe Kotlin is working on improving the interop story for Js and Wasm, so this can likely get simplified in the future.

cedrickcooke commented on core/src/commonTest/kotlin/AutoIncrementKeyObjectStore.kt:
```diff
@@ -1,30 +1,36 @@
-        assertEquals("Username", user.username)
+        assertEquals("Username", user?.username)
+
+        database.close()
+        deleteDatabase("auto-increment-keys")
```
I'm not opposed to losing the test wrapper if there was a real reason it had to go (I'm guessing it didn't play nice with wasm?), but if it's gone there needs to be a `try ... finally` for these cleanup lines. Otherwise a failing test leaves some garbage behind that can make future test runs (which should be successful) fail. Same goes for the other tests

eygraber replied:
There was an issue with Wasm, although I have a few ideas about how to make it work. 

But worst case scenario I will add the `try...finally`

cedrickcooke commented:
Absolutely heroic effort! A few comments, but nothing big/architectural, just some small details.

> There's still weirdness to support common code between js and wasmJs, but hopefully the API surface is a little better now, e.g. a lot of the type checking happen when you create a key instead of validating a key after it was created. There are still footguns available to the user, but they would be there anyways (e.g. unsafe casting).

Love the improvements on this front. The footguns are now, at least, more explicit -- it looks a lot harder to accidentally pass something as `JsAny` than it was as `dynamic`.

mipastgt commented:
Any progress on this front?

eygraber commented:
I've been distracted, but I hope to get back to this in the coming weeks.

twyatt reviewed (commented):
This is a bit outside my wheelhouse, but I'll try to review the code for readability/understandability more so than correctness of interacting with IndexedDB.
Just a couple comments so far. Overall, as @cedrickcooke mentioned, what an amazing effort on this! Thanks!

mipastgt commented on README.md:
```diff
@@ -1,9 +1,10 @@
 # Kotlin IndexedDB
 
-A wrapper around [IndexedDB] which allows for access from Kotlin/JS code using `suspend` blocks and linear, non-callback based control flow.
+A wrapper around [IndexedDB] which allows for access from Kotlin/JS or Kotlin/WasmJs code using `suspend` blocks and linear, non-callback based control flow.
```
I prefer the original wording because it is more precise. The use of IndexedDB is tied to the browser and thus to the wasmJS target and not WASM in general.

eygraber replied:
I did that to differentiate it from wasmWasi, although I guess the fact that it's a browser API should be enough to differentiate implicitly.

eygraber commented on external/src/jsMain/kotlin/Interop.kt:
```diff
@@ -0,0 +1,106 @@
+
+import kotlinx.browser.window
+import org.khronos.webgl.Uint8Array
+import kotlin.js.unsafeCast
+import kotlin.js.unsafeCast as jsUnsafeCast
```
Could be a mistake, I'll double check.

apolostudio commented:
pinging this!

eygraber commented:
Sorry, haven't had much free time to get back to this. Hopefully will be able to soon.

eygraber commented:
Update: real life continues to get in the way of addressing comments :disappointed: 

I haven't forgotten about it, and will come back to it as soon as I'm able, but if anyone wants to pick up where I left off, feel free :smile:

mipastgt commented:
Yet another ping :-)

eygraber commented:
Apologies, I've had non stop work to do since making this PR. Haven't forgotten, but still unsure when I can dedicate time to getting back to it.

eygraber commented:
This may become easier to integrate in Kotlin 2.2.20 because of support for shared source sets between js and wasm - https://kotlinlang.org/docs/whatsnew-eap.html#shared-source-set-for-js-and-wasmjs-targets

mipastgt commented:
Just wanted to add this link to the Kotlin wrappers so that it does not get forgotten.
https://github.com/JetBrains/kotlin-wrappers/tree/master/kotlin-browser/src/webMain/generated/web/idb

benkuly commented:
Be aware that kotlin-wrappers is really API binary unstable. So if you would include it as dependency, you would likely force the users of this lib to use the same version.

N7ghtm4r3 commented:
Hi, any updates about the porting now that Kotlin `2.2.20` is stable? This porting should be very useful!
